File: PriorityQueues/PriorityQueue.cpp
#include <bits/stdc++.h>
using namespace std;

class PriorityQueue{
    vector<int> heap;
    int size;

    public:
        PriorityQueue() {
            heap = vector<int>();
            heap.push_back(0);
            size = 0;
        }

        bool less(int i, int j){
            return i>=1 && i<=size
                && j>=1 && j<=size
                && heap[i] < heap[j];
        }

        void swap(int i, int j){
            if (i!=0 && j!=0) {
                int temp = heap[i];
                heap[i] = heap[j];
                heap[j] = temp;
            }
        }

        void bubbleDown(int i){
            int left = 2*i;
            int child = left + less(left,left+1);
            if (less(i,child)) {
                swap(i,child);
                bubbleDown(child);
            }
        }

        void bubbleUp(int i){
            int parent = i/2;
            if (less(parent,i)){
                swap(parent,i);
                bubbleUp(parent);
            }
        }

        int max(){
            return heap[1];
        }

        void insert(const int &x){
            heap.push_back(x);
            size++;
            bubbleUp(size);
        }

        int extractMax() {
            int max = heap[1];
            heap[1] = heap[size];
            heap.pop_back();
            size--;
            bubbleDown(1);
            return max;
        }

        void print() {
            for (int i=0;i<size;i++){
                cout<<heap[i+1]<<' ';
            }
            cout<<endl;
        }

};

int main() {
    PriorityQueue pq = PriorityQueue();
    int N;
    cin>>N;
    for (int i=0;i<N;i++) {
        char c;
        cin>>c;
        if (c=='I'){
            int x;
            cin>>x;
            pq.insert(x);
        } else if (c=='E'){
            cout<<pq.extractMax()<<endl;
        } else if (c=='P'){
            pq.print();
        }
    }
}

File: Priority Queues/PriorityQueue.py

class PriorityQueue:
    def __init__(self):
        self.heap = [None]
        self.n = 0

    def less(self, x, y):
        return 1<=x<=self.n and 1<=y<=self.n and self.heap[x]<self.heap[y]

    def swap(self, x, y):
        temp = self.heap[x]
        self.heap[x] = self.heap[y]
        self.heap[y] = temp

    def bubbleUp(self,node):
        parent = node//2
        if (self.less(parent,node)):
            self.swap(node,parent)
            self.bubbleUp(parent)

    def bubbleDown(self,node):
        left = node*2
        child = left + self.less(node*2,node*2+1)
        if (self.less(node,child)):
            self.swap(node,child)
            self.bubbleDown(child)

    def max(self):
        return self.heap[1]

    def insert(self, x):
        self.heap.append(x)
        self.n += 1
        self.bubbleUp(self.n)

    def extractMax(self):
        r = self.heap[1]
        self.heap[1] = self.heap[self.n]
        self.heap.pop()
        self.n -= 1
        self.bubbleDown(1)
        return r
    
    def increaseKey(self, x, key):
        self.heap[x] = key
        self.bubbleUp[x]

    def print(self):
        for i in range(self.n):
            print(self.heap[i+1],end=' ')
        print()

N = int(input())
lines = [input().split() for _ in range(N)]

pq = PriorityQueue()
for line in lines:
    if line[0]=='I':
        pq.insert(int(line[1]))
    elif line[0]=='E':
        print(pq.extractMax())
    elif line[0]=='P':
        pq.print()
